spec/helpers/barnardos/form_helpers_spec.rb

```diff
@@ -0,0 +1,130 @@
+require 'rails_helper'
+
+RSpec.describe Barnardos::ActionView::FormHelpers, :type => :helper do
+  include RSpecHtmlMatchers
+
+  describe '#labelled_text_field_tag' do
+    ##
+    # <!-- Example HTML output -->
+    # <div class="textfield " id="participant_name-wrapper">
+    #   <label class="textfield__label textfield__label--bold" for="participant_name">
+    #     What is the name of the research participant?
+    #     <span class="textfield__hint">Full name</span>
+    #   </label>
+    #   <input id="participant_name" class="textfield__input" name="participant_name" type="text">
+    # </div>
+    context 'with a name and a label parameter' do
+      let(:name) { :participant_name }
+      let(:label) { 'What is the name of the research participant?' }
+
+      subject(:rendered) { helper.labelled_text_field_tag(name, label) }
+
+      it 'outputs an wrapper div with the class and id' do
+        expect(rendered).to have_tag('div', with: { class: 'textfield', id: 'participant_name-wrapper' })
+      end
+
+      it 'renders a classed bold label and input as sibling children of the div' do
```

Comment:
There are three tests:

```ruby
      it 'outputs an wrapper div with the class and id' do
        expect(rendered).to have_tag('div', with: { class: 'textfield', id: 'participant_name-wrapper' })
      end

      it 'renders a classed bold label and input as sibling children of the div' do
        expect(rendered).to have_tag('div.textfield > label.textfield__label.textfield__label--bold', text: label)
        expect(rendered).to have_tag('div.textfield > input.textfield__input')
        expect(rendered).to have_tag('label[for=participant_name]')
      end

      it 'names the input as given' do
        expect(rendered).to have_tag('input', with: { name: 'participant_name', type: 'text' })
      end
```

We *could* have one test which fails if *anything* is wrong, and this is something I've done in the past. Why am I not doing it here? Independent failure. If we wrote one test the `it` would have to read "Render a labelled input box inside a wrapper div that's all classed correctly`. You can do that, but if it fails you have to parse everything that *might* be wrong, and the most important thing in terms of a working site is that the input is named correctly (which is why it's got its own test – I'm less sure about the wrapper `div`).

Ordering is a harder problem in pure RSpec, there aren't any explicit before/after matchers for that kind of thing, so you need to fall back on `first-child`, `last-child`, `nth-child` in the CSS selectors (could add those in the middle test).

You can argue, probably quite successfully, that I've split these up badly. Do you fancy having a go at making it better?

You're right about the specs that need to follow. We'll need new `context`s for these.